frontend/src/context/Web3Context.tsx

```diff
@@ -0,0 +1,95 @@
+import React, {
+  createContext,
+  useContext,
+  useEffect,
+  useState,
+  useCallback,
+} from "react";
+import Web3 from "web3";
+import { contractABI } from "../lib/abi";
+import { magic } from "../lib/magic";
+
+// Define the structure of the Web3 context state
+type Web3ContextType = {
+  web3: Web3 | null;
+  initializeWeb3: () => void;
+  contract: any;
+  isAccountChanged: boolean;
+};
+
+// Define contract address
+const CONTRACT_ADDRESS = "0xf4759a2bf9a8b6dc8318efc53e6e27b452c42310";
+
+// Create the context with default values
+const Web3Context = createContext<Web3ContextType>({
+  web3: null,
+  initializeWeb3: () => {},
+  contract: null,
+  isAccountChanged: false,
+});
+
+// Custom hook to use the Web3 context
+export const useWeb3 = () => useContext(Web3Context);
+
+// Provider component to wrap around components that need access to the context
+export const Web3Provider = ({ children }: { children: React.ReactNode }) => {
+  // State variable to hold an instance of Web3 and the contract
+  const [web3, setWeb3] = useState<Web3 | null>(null);
+  const [contract, setContract] = useState<any | null>(null);
+  const [isAccountChanged, setIsAccountChanged] = useState<boolean>(false);
+
+  // Initialize Web3
+  const initializeWeb3 = useCallback(async () => {
+    try {
+      // Get the provider from the Magic instance
+      const provider = await magic.wallet.getProvider();
+
+      // Create a new instance of Web3 with the provider
+      const web3Instance = new Web3(provider);
+
+      // Subscribe to accounts changed event
+      provider.on("accountsChanged", async () => {
+        setIsAccountChanged((state) => !state);
+      });
+
+      // Subscribe to chain changed event
+      provider.on("chainChanged", async () => {
+        const chainId = await web3Instance.eth.getChainId();
+        const sepoliaChainId = 11155111;
+        if (chainId !== sepoliaChainId) {
+          alert("Please switch to the Sepolia network");
+        }
+      });
```

Comment:
> I didn't see any handling for events that would be emitted by third party wallets (chainChanged, accountsChanged) - if a user connects with metamask and then switches their wallet, is this app handling that in real time? Or is the user required to reload the page in order for the app to pull in the new account?

Updated with chainChanged and accountsChanged. For accountsChanged, it will trigger `fetchData` to run in the UserContext to update the user state.